File: Array and Lists/two_sum.py
# Link: https://leetcode.com/problems/two-sum/

# Approach: Add all the numbers in the dictionary with corresponding index in the array. Now again start iterating over elements in the array. Perform num = target-array[i].
# If the num is in dictionary then extract its position (let's say j) and return (i, j).

class Solution(object):
    def twoSum(self, nums, target):
        dic = {}
        
        for i in range(len(nums)):
            dic[nums[i]] = i
            
        for i in range(len(nums)):
            if target-nums[i] in dic:
                j = dic[target-nums[i]]
                if i != j:
                    return [min(i, j), max(i, j)]

File: Array and Lists/two_sum.cc
// Link: https://leetcode.com/problems/two-sum/

class Solution 
{
    public:
        vector<int> twoSum(vector<int>& nums, int target) 
        {
            map<int, int> values;
            map<int,int>::iterator it;
            vector<int> ans;
            
            int temp;
            
            for(int i=0;i<nums.size();i++)
            {
                values.insert(make_pair(nums[i], i));
            }
            
            for(int i=0;i<nums.size();i++)
            {
                temp = target - nums[i];
                
                it = values.find(temp);
                if(it != values.end())
                {
                    if(values[temp] != i)
                    {
                        if(values[temp]>i)
                        {
                            ans.push_back(i);
                            ans.push_back(values[temp]);
                            return ans;
                        }
                        else
                        {
                            ans.push_back(values[temp]);
                            ans.push_back(i);
                            return ans;
                        }
                    }
                }
            }
            return ans;
        }
}; 
